File: Sorting/HeapSort.java
public class HeapSort
{
	int ar[];
	public static void main(String args[])
	{
		HeapSort hs = new HeapSort();
		hs.ar = new int[args.length];
		for(int i=0; i<args.length;i++)
		{
			hs.ar[i]=Integer.parseInt(args[i]);
		}
		hs.heapify();
		hs.printArray();
		hs.heapSort();	
		hs.printArray();
	}

	public void printArray()
	{
		for(int i:ar)
		{
			System.out.println(i);
		}
		
	}
	public void heapSort()
	{
		for(int i=0;i<ar.length;i++)
		{
			swap(0,ar.length-1-i);
			bubbleDown(0,ar.length-2-i);
		}

	}
	public void heapify()
	{
		for(int i=0;i<ar.length;i++)
		{
			bubbleUp(i);
		}
	}
	
	public void bubbleDown(int i,int lastIndex)
	{
		while(i<lastIndex && (i*2)+1<lastIndex)
		{
			int child1= i*2+1;
			int child2= i*2+2;
			int swapIndex = i;
			if(child1<lastIndex && child2<lastIndex)
			{
				swapIndex = ar[child1]>ar[i]?child1:i;
				swapIndex = ar[child2]>ar[swapIndex]?child2:swapIndex;
				if(i!=swapIndex)
				{
					swap(i,swapIndex);
					i=swapIndex;
				}
				else
				{
					break;	
				}
			}
			else
			{
				if(child1<lastIndex)
				{
					if(ar[child1]>ar[i])
					{
						swap(child1,i);
						i= child1;
					}
					else
					{
						break;
					}
				}
			}
	
		}	
	
	}
	
	public void bubbleUp(int i)
	{
		while(i>=0)
		{
			int parent = (i-1)/2;
			if(parent>=0)
			{
				if(ar[parent]<ar[i])
				{
					swap(parent,i);
					i=parent;
				}
				else
				{
					break;
				}

			}
			else
			{
				break;
			}

		}

	}
	public void swap(int a , int b)
	{
		int temp = ar[a];
		ar[a]= ar[b];
		ar[b]= temp;
	}
	
}

File: Arrays/heapSort.py

ar = [4,5,3,8,9,123,34,65,9032]

def heapify(ar):
	for i in range(len(ar)/2,-1,-1):
		bubbleDown(ar,i,len(ar)-1)
	return ar

def heapSort(ar):
	print  heapify(ar)
	lastIndex = len(ar)-1
	while lastIndex!=0:
		swap(ar,0,lastIndex)
		lastIndex-=1
		bubbleDown(ar,0,lastIndex)
	return ar

def bubbleDown(ar,i,maxIndex):
	while i<maxIndex:
		child1 = i*2+1
		child2 = i*2+2
		swapindex =i
		if child1>maxIndex:
			break
		else:
			if ar[child1]>ar[i]:
				swapindex = child1
			if child2<=maxIndex:
				if ar[child2]>ar[swapindex]:
					swapindex=child2
			if i == swapindex:
				break
			else:
				swap(ar,swapindex,i)
				i= swapindex
	print ar
	
def bubbleUp(ar,index):
	while(index!=0):
		parentIndex = (index-1)/2
		if ar[parentIndex]<ar[index]:
			swap(ar,index,parentIndex)
			index = parentIndex
		else:
			break
	return ar

def swap(ar,index1, index2):
	temp = ar[index1]
	ar[index1] = ar[index2]
	ar[index2] = temp

def heapify2(ar,i,lastIndex):
	if i>=lastIndex:
		return
	else:
		swapindex=i
		child1=  i*2+1
		child2 = i*2+2
		if child1<=lastIndex:
			if ar[child1]>ar[i]:
				swapindex= child1	
			if child2<=lastIndex:
				if ar[swapindex]<ar[child2]:
					swapindex = child2
			if swapindex!=i:
				swap(ar,i,swapindex)
				heapify2(ar,swapindex,lastIndex)
	return ar

ar = [4,5,3,8,9,123,34,65,9032]
print heapSort(ar)

